File: go/tree/inorderTraversal.go
package tree

func inorderTraversal(root *TreeNode) []int {
	var path []int
	var stack []*TreeNode
	curr := root
	for curr != nil || len(stack) != 0 { // 左 父 右
		for curr != nil {

			stack = append(stack, curr)
			curr = curr.Left
		}
		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		path = append(path, curr.Val) // in
		curr = curr.Right
	}
	return path
}

File: java/src/main/java/io/dure/tree/InorderTraversal.java
package io.dure.tree;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class InorderTraversal {

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> path = new ArrayList<>();
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            path.add(curr.val);
            curr = curr.right;
        }
        return path;
    }
}
